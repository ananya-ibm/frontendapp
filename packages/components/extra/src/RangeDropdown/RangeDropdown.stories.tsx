/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React from 'react';
import { Slider } from '@exo/frontend-components-base';
import { RangeDropdown } from './RangeDropdown';

export default {
  title: 'Components/Extra/RangeDropdown',
  component: RangeDropdown
};

const options = [
  {
    label: 'How much will you pay for your deposit?',
    min: 0,
    max: 30000,
    step: 1000,
    value: 3000,
    unit: '£'
  },
  {
    label: 'How many miles per annum?',
    min: 1000,
    max: 30000,
    step: 1000,
    value: 10000,
    unit: '£'
  },
  {
    label: 'How many months would you like to pay over?',
    min: 12,
    max: 48,
    step: 6,
    value: 36,
    unit: 'Month'
  }
];

const SliderChildren = () => {
  const onChange = (idx, evt) => {
    // eslint-disable-next-line no-console
    console.log('Slider ', idx, ' clicked with ', evt.value);
  };

  return (
    <>
      {options.map((option, idx) => (
        <Slider
          key={option.label}
          labelText={option.label}
          min={option.min}
          max={option.max}
          step={option.step}
          value={option.value}
          onChange={evt => onChange(idx, evt)}
          minLabel={option.unit}
          maxLabel={option.unit}
        />
      ))}
    </>
  );
};

const storyProps = {
  label: 'Budget',
  children: <SliderChildren />
};

export const normal = args => <RangeDropdown {...args} />;
normal.args = storyProps;
