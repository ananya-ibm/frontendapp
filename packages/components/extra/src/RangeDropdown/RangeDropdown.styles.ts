/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import styled from 'styled-components';

export const RangeDropdown = styled('div')`
  position: relative;

  & .cds--dropdown {
    width: 8rem;
  }
`;

export const RangeContainer = styled('div')`
  background-color: white;
  box-shadow: 0 0.2rem 0.6rem rgba(0, 0, 0, 0.2);
  max-width: 40rem;
  overflow-y: auto;
  position: absolute;
  transition: max-height 110ms cubic-bezier(0.2, 0, 0.38, 0.9);
  z-index: 11;

  &:focus {
    outline: 0.1rem solid #0f62fe;
  }
`;

export const RangeChildren = styled('div')`
  padding: 2rem;

  & .cds--slider-text-input {
    width: unset;
  }
`;
