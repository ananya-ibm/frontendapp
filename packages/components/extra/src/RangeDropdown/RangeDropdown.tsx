/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React, { useEffect, useState } from 'react';
import { ChevronDown, ChevronUp } from '@carbon/react/icons';
import * as S from './RangeDropdown.styles';

export const RangeDropdown = ({ label = 'Budget', isExpanded = false, children }: Props) => {
  const [expanded, setExpanded] = useState(isExpanded);
  const componentRef = React.createRef<HTMLDivElement>();
  const containerRef = React.createRef<HTMLDivElement>();

  const onClickOutside = evt => {
    // ignore clicks on the component itself
    if (componentRef.current?.contains(evt.target)) {
      return;
    }
    setExpanded(false);
  };

  useEffect(() => {
    document.addEventListener('click', onClickOutside, false);
    return () => {
      document.removeEventListener('click', onClickOutside, false);
    };
  });

  const onClick = () => {
    setExpanded(!expanded);
    containerRef.current?.focus();
  };

  return (
    <S.RangeDropdown ref={componentRef}>
      <div
        className={
          expanded
            ? 'cds--dropdown cds--list-box cds--dropdown--light cds--list-box--light cds--dropdown--xl cds--list-box--xl cds--dropdown--open cds--list-box--expanded'
            : 'cds--dropdown cds--list-box cds--dropdown--light cds--list-box--light cds--dropdown--xl cds--list-box--xl'
        }
      >
        <button
          type="button"
          className="cds--list-box__field"
          aria-haspopup="listbox"
          aria-expanded={expanded}
          onClick={onClick}
        >
          <span className="cds--list-box__label">{label}</span>
          <div className="cds--list-box__menu-icon">
            {expanded ? <ChevronUp size={16} /> : <ChevronDown size={16} />}
          </div>
        </button>
      </div>

      <S.RangeContainer
        role="listbox"
        ref={containerRef}
        aria-label="Range Container"
        title="range container"
      >
        {expanded && <S.RangeChildren>{children}</S.RangeChildren>}
      </S.RangeContainer>
    </S.RangeDropdown>
  );
};

type Props = {
  children?: any;
  label: string;
  className?: string;
  isExpanded?: boolean;
};
