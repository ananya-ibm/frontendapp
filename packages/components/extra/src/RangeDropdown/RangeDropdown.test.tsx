/*
Licensed Materials - Property of IBM
694906H
(c) Copyright IBM Corp.  2020 All Rights Reserved

US Government Users Restricted Rights - Use, duplication or disclosure restricted
by GSA ADP Schedule Contract with IBM Corp.
*/

import React from 'react';
import { renderWithTheme as render } from '@testUtils';
import { fireEvent } from '@testing-library/react';
import { axe } from 'jest-axe';
import { RangeDropdown } from './RangeDropdown';

const testProps = {
  label: 'test',
  children: <div>children</div>
};

describe('<RangeDropdown /> component', () => {
  test('its snapshot matches', () => {
    const { container } = render(<RangeDropdown {...testProps} />);

    expect(container).toBeDefined();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should open when clicked', async () => {
    const { getByText, findByText } = render(<RangeDropdown {...testProps} />);
    const openText = getByText(testProps.label);
    expect(openText).toBeDefined();
    fireEvent.click(openText);

    const expandedText = await findByText('children');
    expect(expandedText).toBeDefined();
  });

  test('it passes an axe accesibility check', async () => {
    const { container } = render(<RangeDropdown {...testProps} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
