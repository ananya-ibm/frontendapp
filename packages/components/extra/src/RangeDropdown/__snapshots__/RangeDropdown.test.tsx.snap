// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeDropdown /> component its snapshot matches 1`] = `
.c0 {
  position: relative;
}

.c0 .cds--dropdown {
  width: 8rem;
}

.c1 {
  background-color: white;
  box-shadow: 0 0.2rem 0.6rem rgba(0,0,0,0.2);
  max-width: 40rem;
  overflow-y: auto;
  position: absolute;
  -webkit-transition: max-height 110ms cubic-bezier(0.2,0,0.38,0.9);
  transition: max-height 110ms cubic-bezier(0.2,0,0.38,0.9);
  z-index: 11;
}

.c1:focus {
  outline: 0.1rem solid #0f62fe;
}

<div
  class="c0"
>
  <div
    class="cds--dropdown cds--list-box cds--dropdown--light cds--list-box--light cds--dropdown--xl cds--list-box--xl"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="cds--list-box__field"
      type="button"
    >
      <span
        class="cds--list-box__label"
      >
        test
      </span>
      <div
        class="cds--list-box__menu-icon"
      >
        <svg
          aria-label="icon"
          fill="currentColor"
          focusable="false"
          height="20"
          preserveAspectRatio="xMidYMid meet"
          role="img"
          size="16"
          viewBox="0 0 32 32"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        />
      </div>
    </button>
  </div>
  <div
    aria-label="Range Container"
    class="c1"
    role="listbox"
    title="range container"
  />
</div>
`;
